File: 이하영/240612/2250. 트리의 높이와 너비/트리의 높이와 너비.py
# 각 레벨의 너비 = 맨 오른쪽 노드 열 번호 - 맨 왼쪽 노드 열 번호 + 1
# 레벨은 1(루트노드)부터 1씩 증가하며 내려감
# 구하는 것 : 너비가 가장 넓은 레벨, 그 레벨의 너비 ( 단, 정답이 될 레벨이 2개 이상이라면 가장 작은 값의 레벨을 답으로)
# 노드의 위치 => 구하려는 노드의 왼쪽에 위치하는 노드의 개수 + 1 -> 중위 순회로 
# 트리 노드 -> 부모 노드가 없는 노드
# 트리 깊이 -> DFS
# 노드 저장 tree[i][0] = 왼, tree[i][1] = 오
import sys
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

N = int(input()) # 노드의 개수
tree = [[-1, -1] for _ in range(N+1)] # -1은 노드가 들어있지 않다는 뜻
tree_parent = [-1] * (N+1)

column = 1
node_info = dict() # key : 노드 깊이, value : 노드 번호, 노드 열

def depth_and_col (v, depth): #노드 번호, 깊이
    global column #열은 겹치지 않기 때문에
    
    if v == -1: # 노드가 들어있지 않으면 리턴
        return
    
    depth_and_col(tree[v][0], depth+1) #왼쪽
    
    if depth not in node_info: # 새로운 층이라면
        node_info[depth] = [] 
    node_info[depth].append((v, column)) # 루트
    column += 1
    
    depth_and_col(tree[v][1], depth+1) # 오른쪽
    
        
for _ in range(N):
    me, l, r = map(int, input().split())
    tree[me][0] = l
    tree[me][1] = r
    
    if l != -1 :
        tree_parent[l] = me
    if r != -1 :
        tree_parent[r] = me

for i in range(1, N+1):
    if tree_parent[i] == -1 and(tree[i]): #부모 노드는 없고, 자식 노드는 있을 때
        root = i
        break

depth_and_col(root, 1)

max_width = 0
max_level = 0

for level in sorted(node_info.keys()):
    cols = [info[1] for info in node_info[level]] # info[0] = 노드 번호, info[1] = 레벨
    
    width = max(cols) - min(cols) + 1# 해당 레벨에서 가장 큰 열 - 가장 작은 열
    if width > max_width: # 저장된 너비보다 이번 레벨의 너비가 더 크다면
        max_width = width
        max_level = level
        
print(max_level, max_width)

File: 한민선/240612/2250. 트리의 높이와 너비/트리의 높이와 너비.java
import java.io.*;
import java.util.*;

public class Main {
    static class node{
        int left;
        int right;
        int parent;
        node(int left, int right){
            this.left = left;
            this.right = right;
            this.parent = -1;
        }
    }
    // 깊이 별 노드 저장 -> 노드의 번호 말고 <레벨, 열 번호> 로 저장
    static HashMap<Integer, ArrayList<Integer>> levels;
    // 각 레벨의 노드의 열 번호 저장할 변수
    static int column;
    static node[] tree;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        // 1. 트리 만들기
        // 2. DFS 를 이용하여 깊이와 각 레벨 별 노드의 열 번호 구하기
        // - 키, 값 형식으로 저장
        // 3. 너비의 최댓값 비교해서 구하기
        // - 너비 = 맨 오른쪽 노드(max) - 맨 왼쪽 노드(min) + 1

        int n = Integer.parseInt(br.readLine());
        tree = new node[n+1];
        for (int i = 0; i < n+1; i++) {
            tree[i] = new node(-1, -1);
        }
        int root = 0;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int v = Integer.parseInt(st.nextToken());
            int l = Integer.parseInt(st.nextToken());
            int r = Integer.parseInt(st.nextToken());

            tree[v].left = l;
            tree[v].right = r;

            if(tree[v].left != -1)
                tree[l].parent = v;

            if(tree[v].right != -1)
                tree[r].parent = v;
        }

        for (int i = 1; i < n+1; i++) {
            // 부모 노드가 없으면 루트노드
            if(tree[i].parent == -1){
                root = i;
                break;
            }
        }

        levels = new HashMap<>();
        int depth = 0;
        column = 0;
        // 중위 순회
        dfs(root, depth);

        int maxWid = 0;
        int maxLv = 0;

        for(int key : levels.keySet()){
            // 레벨의 열 정보들
            Integer[] cols = levels.get(key).stream().toArray(Integer[]::new);
            Arrays.stream(cols).sorted();
            // 너비 = max - min + 1
            int w = cols[cols.length - 1] - cols[0] + 1;
            if(w > maxWid){
                maxWid = w;
                maxLv = key;
            }
        }

        bw.write(maxLv + 1 + " " + maxWid);
        bw.flush();
        bw.close();
    }

    // 깊이 별 열구하기
    static void dfs(int idx, int d){
        if(idx == -1){
            return;
        }
        node n = tree[idx];

        if(n == null) return;

        // 해당 레벨이 없으면 해시맵 생성
        if(!levels.containsKey(d)){
            levels.put(d, new ArrayList<>());
        }

        dfs(n.left, d + 1);

        column += 1;
        levels.get(d).add(column);

        dfs(n.right, d + 1);
    }
}